src/main.rs

```diff
@@ -221,101 +229,100 @@ async fn process_created(path: PathBuf, uid: u32, gid: u32, workingdir: PathBuf)
-    let re = Regex::new("/dev/sd[a-zA-Z]{1,2}[0-9]{1,2}").unwrap();
-    if re.find(path.as_os_str().to_str().unwrap_or("")).is_none() {
-        return;
-    };
+async fn process_removed(path: PathBuf, workingdir: PathBuf) -> anyhow::Result<()> {
+  let re = Regex::new("/dev/sd[a-zA-Z]{1,2}[0-9]{1,2}").unwrap();
+  if re.find(path.as_os_str().to_str().unwrap_or("")).is_none() {
+    return Ok(());
+  };
 
-    guard!(let Some(device) = extract_device(path.as_path()).await else {
-      tracing::warn!("couldn't match the device for some reason?");
-      return;
-    });
-    tracing::info!("processing {:?}", path);
+  guard!(let Some(device) = extract_device(path.as_path()).await else {
+    tracing::warn!("couldn't match the device for some reason?");
+    return Ok(());
+  });
 
-    let dest: std::path::PathBuf = workingdir.join(PathBuf::from(device));
-    tracing::debug!("destpath is {:?}", dest);
+  tracing::info!("processing {:?}, event=removed", path);
+  notify(format!("removed device: {}", device).as_str())?;
 
-    if let Err(why) = unmount(&dest, UnmountFlags::FORCE) {
-        tracing::error!("failed to unmount {:?}: {}", path, why);
-    }
+  let dest: std::path::PathBuf = workingdir.join(PathBuf::from(device));
+  tracing::debug!("destpath is {:?}", dest);
+
+  if let Err(why) = unmount(&dest, UnmountFlags::FORCE) {
+    tracing::error!("failed to unmount {:?}: {}", path, why);
+  }
 
-    drop(remove_dir(dest).await);
+  drop(remove_dir(dest).await);
+  Ok(())
 }
 
 /// Yo, main body what do you expect. It does whatever usually main does.
 #[tokio::main]
 async fn main() -> anyhow::Result<()> {
-    // ### logging setup
-    tracing::info!("\n\nset DEBUGMODE to anything for easier debugging\n\n");
-    let is_debug_mode = std::env::var("DEBUGMODE").map_or(false, |x| !x.is_empty());
-
-    let subscriber = tracing_subscriber::registry().with(tracing_subscriber::EnvFilter::new(
-        std::env::var("RUST_LOG").unwrap_or_else(|_| "debug".into()),
-    ));
-    if is_debug_mode {
-        subscriber
-            .with(
-                tracing_subscriber::fmt::layer()
-                    .with_line_number(true)
-                    .with_thread_ids(true)
-                    .with_thread_names(true)
-                    .pretty(),
-            )
-            .init()
-    } else {
-        subscriber
-            .with(tracing_subscriber::fmt::layer().compact())
-            .init();
-    }
+  // ### logging setup
+  tracing::info!("\n\nset DEBUGMODE to anything for easier debugging\n\n");
+  let is_debug_mode = std::env::var("DEBUGMODE").map_or(false, |x| !x.is_empty());
 
-    // ### preparations
-    let settings = Config::builder()
-        .add_source(config::Environment::with_prefix("RUDEVIL").separator("_"))
-        .build()
-        .unwrap();
+  let subscriber = tracing_subscriber::registry().with(tracing_subscriber::EnvFilter::new(
+    std::env::var("RUST_LOG").unwrap_or_else(|_| "debug".into()),
```

Comment:
You should extract this to another function

Reply:
actually, yeah its global allocation isn't it ?

Reply:
fixed